zio-http/shared/src/main/scala/zio/http/internal/QueryParamEncoding.scala

```diff
@@ -17,12 +17,311 @@
+      var startIndex = 0
+
+      // Skip a leading '?' if present
+      if (startIndex < length && queryStringFragment.charAt(startIndex) == '?') {
+        startIndex += 1
+      }
+
+      while (startIndex < length) {
+        // Find key-value separator ('=') or parameter separator ('&')
+        var eqIndex  = -1
+        var ampIndex = queryStringFragment.indexOf('&', startIndex)
+        if (ampIndex == -1) ampIndex = length
+
+        // Find equals sign within current parameter segment
+        var i = startIndex
+        while (i < ampIndex && eqIndex == -1) {
+          if (queryStringFragment.charAt(i) == '=') eqIndex = i
+          i += 1
+        }
+
+        // Extract and decode key and value directly using indices
+        var key = if (eqIndex != -1) {
+          decodeComponent(queryStringFragment, startIndex, eqIndex, charset)
+        } else {
+          decodeComponent(queryStringFragment, startIndex, ampIndex, charset)
+        }
+
+        var value = if (eqIndex != -1) {
+          decodeComponent(queryStringFragment, eqIndex + 1, ampIndex, charset)
+        } else {
+          ""
+        }
+
+        // This might seem strange. However, we copied this behavior from Netty.
+        if (key == "") {
+          key = value
+          value = ""
+        }
+
+        // Add the parameter to our map
+        // Get or create values list for this key without allocating a lambda
+        var values = params.get(key)
+        if (values == null) {
+          // Key doesn't exist yet, create new list
+          values = new util.ArrayList[String](1)
+          params.put(key, values)
+        }
+        values.add(value)
+
+        // Move to the next parameter
+        startIndex = ampIndex + 1
+      }
+
+      QueryParams(params)
+    }
+  }
+
+  def encode(baseUri: StringBuilder, queryParams: QueryParams, charset: Charset): String = {
+    if (queryParams.isEmpty) {
+      return baseUri.toString
+    }
+
+    // Estimate additional capacity needed for query parameters
+    // Initial guess: 1 char for ? + average of 20 chars per param pair
+    val additionalCapacity = 1 + (queryParams.seq.size * 20)
+    baseUri.ensureCapacity(baseUri.length + additionalCapacity)
+
+    var isFirst  = true
+    val iterator = queryParams.seq.iterator
+    while (iterator.hasNext) {
+      val entry  = iterator.next()
+      val key    = entry.getKey
+      val values = entry.getValue
+
+      if (key != "") {
+        if (values.isEmpty) {
+          // Handle key with no value
+          if (isFirst) {
+            baseUri.append('?')
+            isFirst = false
+          } else {
+            baseUri.append('&')
+          }
+          baseUri.append(encodeComponent(key, charset))
+          baseUri.append('=')
+        } else {
+          // Handle key with one or more values
+          import scala.jdk.CollectionConverters._
+          values.asScala.foreach { value =>
```

Comment:
I think you can use an iterator on `values` with a while loop

Reply:
Used while with index access. This is a Java list, most likely ArrayList. So get is O(1) and I save any kind of allocation.